File: jskat-gui/src/main/java/org/jskat/gui/swing/table/AbstractHandPanel.java
package org.jskat.gui.swing.table;

import org.jskat.data.JSkatOptions;
import org.jskat.gui.img.JSkatGraphicRepository;
import org.jskat.gui.swing.LayoutFactory;
import org.jskat.util.*;

import javax.swing.*;
import javax.swing.border.Border;
import java.awt.*;

/**
 * Abstract class for a panel representing a players hand
 */
abstract class AbstractHandPanel extends JPanel {


    /**
     * Player position
     */
    Player position;
    /**
     * Card images
     */
    JSkatGraphicRepository bitmaps;
    /**
     * i18n strings
     */
    JSkatResourceBundle strings;
    /**
     * Header panel
     */
    JPanel header;
    /**
     * Header label
     */
    JLabel headerLabel;
    /**
     * Icon panel
     */
    IconPanel iconPanel;
    /**
     * Clock panel
     */
    ClockPanel clockPanel;
    /**
     * Player name
     */
    String playerName;
    /**
     * Player bid
     */
    int bidValue;

    boolean showIssWidgets;

    ClickableCardPanel cardPanel;

    /**
     * Maximum card count
     */
    int maxCardCount = 0;

    /**
     * Indicates whether this player is currently the active player
     */
    boolean isActivePlayer = false;

    /**
     * Indicates whether this player is an AI player
     */
    boolean isAIPlayer = false;

    boolean playerPassed = false;

    boolean playerGeschoben = false;

    boolean declarer = false;

    private boolean playerContra;

    private boolean playerRe;

    /**
     * Constructor
     *
     * @param actions        Action map
     * @param maxCards       the maximum number of cards
     * @param showIssWidgets TRUE, if ISS widgets should be shown
     */
    AbstractHandPanel(final ActionMap actions, final int maxCards, final boolean showIssWidgets) {

        setActionMap(actions);
        this.bitmaps = JSkatGraphicRepository.INSTANCE;
        this.strings = JSkatResourceBundle.INSTANCE;
        this.maxCardCount = maxCards;
        this.showIssWidgets = showIssWidgets;

        setBackground(new Color(72, 103, 96));

        this.headerLabel = new JLabel(" ");
        this.iconPanel = new IconPanel();
        this.clockPanel = new ClockPanel();

        initPanel();
    }

    /**
     * Initializes the panel
     */
    void initPanel() {

        setLayout(LayoutFactory.getMigLayout("fill, insets 0", "fill", "[shrink][grow]"));

        setBorder(getPanelBorder(this.isActivePlayer));

        String headerInsets = "insets ";
        if (showIssWidgets) {
            headerInsets = headerInsets + "0 5 0 0";
        } else {
            headerInsets = headerInsets + "5";
        }

        this.header = new JPanel(LayoutFactory.getMigLayout("fill, " + headerInsets, "[shrink][grow][shrink]", "fill"));
        this.header.add(this.headerLabel);
        // blank panel
        this.header.add(new JPanel());
        this.header.add(this.iconPanel);
        if (this.showIssWidgets) {
            iconPanel.setShowIssWidgets(true);
            this.header.add(this.clockPanel);
        }
        add(this.header, "shrinky, wrap");

        this.cardPanel = new ClickableCardPanel(this, 1.0, true);
        add(this.cardPanel, "growy");

        if (JSkatOptions.instance().isCheatDebugMode().booleanValue()) {
            showCards();
        }
    }

    /**
     * Retrieves the Border for the hand panel for that player. It visually
     * distinguishes between active and non-active players.
     *
     * @return
     */
    private Border getPanelBorder(final boolean isActivePlayer) {

        Border resultingBorder = null;

        if (isActivePlayer) {
            resultingBorder = BorderFactory.createLineBorder(new Color(255, 191, 0), 3);
        } else {
            resultingBorder = BorderFactory.createLineBorder(Color.black, 3);
        }

        return resultingBorder;
    }

    /**
     * Sets the player position
     *
     * @param newPosition Position
     */
    void setPosition(final Player newPosition) {

        this.position = newPosition;
        refreshHeader();
    }

    void setBidValue(final int newBidValue) {

        this.bidValue = newBidValue;
        refreshHeader();
    }

    /**
     * Gets the player position
     *
     * @return Player position
     */
    Player getPosition() {

        return this.position;
    }

    /**
     * Refreshes the header of this hand panel, i.e. the text and the thinking icon.
     */
    private void refreshHeader() {

        final StringBuffer headerText = new StringBuffer();

        headerText.append(this.playerName).append(": ");

        if (this.position != null) {
            switch (this.position) {
                case FOREHAND:
                    headerText.append(this.strings.getString("forehand"));
                    break;
                case MIDDLEHAND:
                    headerText.append(this.strings.getString("middlehand"));
                    break;
                case REARHAND:
                    headerText.append(this.strings.getString("rearhand"));
                    break;
            }

            headerText.append(" " + this.strings.getString("bid") + ": ");
            headerText.append(this.bidValue);

            if (playerPassed || playerGeschoben || playerContra || playerRe) {

                headerText.append(" (");

                String passedGeschobenContraRe = "";
                if (playerPassed) {
                    passedGeschobenContraRe = strings.getString("passed");
                }
                if (playerGeschoben) {
                    passedGeschobenContraRe = strings.getString("geschoben");
                }

                if (passedGeschobenContraRe.length() > 0 && (playerContra || playerRe)) {
                    passedGeschobenContraRe += " ";
                }

                if (playerContra) {
                    passedGeschobenContraRe += strings.getString("contra");
                }
                if (playerRe) {
                    passedGeschobenContraRe += strings.getString("re");
                }

                headerText.append(passedGeschobenContraRe);

                headerText.append(")");
            }

            if (this.declarer) {
                headerText.append(" (" + this.strings.getString("declarer") + ")");
                var font = headerLabel.getFont();
                headerLabel.setFont(font.deriveFont(font.getStyle() | Font.BOLD));
            } else {
                var font = headerLabel.getFont();
                headerLabel.setFont(font.deriveFont(font.getStyle() & ~Font.BOLD));
            }
        }

        iconPanel.setThinking(isActivePlayer && isAIPlayer);

        this.headerLabel.setText(headerText.toString());
    }

    /**
     * Adds a card to the panel
     *
     * @param newCard Card
     */
    void addCard(final Card newCard) {

        this.cardPanel.addCard(newCard);
    }

    /**
     * Adds a Collection of cards to the panel
     *
     * @param newCards card collection
     */
    void addCards(final CardList newCards) {

        this.cardPanel.addCards(newCards);
    }

    /**
     * Removes a card from the panel
     *
     * @param cardToRemove Card
     */
    void removeCard(final Card cardToRemove) {

        this.cardPanel.removeCard(cardToRemove);
    }

    /**
     * Removes all cards from the panel
     */
    void removeAllCards() {

        this.cardPanel.clearCards();
    }

    /**
     * Removes all cards from the panel and resets other values
     */
    void clearHandPanel() {

        this.cardPanel.clearCards();
        this.bidValue = 0;
        this.playerPassed = false;
        this.playerGeschoben = false;
        this.playerContra = false;
        this.playerRe = false;
        this.declarer = false;
        this.iconPanel.reset();
        refreshHeader();
        setActivePlayer(false);
        hideCards();
    }

    /**
     * Hides all cards on the panel
     */
    void hideCards() {

        this.cardPanel.hideCards();
    }

    /**
     * Shows all cards on the panel
     */
    void showCards() {

        this.cardPanel.showCards();
    }

    void setSortGameType(final GameType newGameType) {

        this.cardPanel.setSortType(newGameType);
    }

    boolean isHandFull() {

        return this.cardPanel.getCardCount() == this.maxCardCount;
    }

    public void setPlayerName(final String newName) {

        this.playerName = newName;

        refreshHeader();
    }

    void setPlayerTime(final double newTime) {

        this.clockPanel.setPlayerTime(newTime);
    }

    void setChatEnabled(final boolean isChatEnabled) {

        this.iconPanel.setChatEnabled(isChatEnabled);
    }

    void setReadyToPlay(final boolean isReadyToPlay) {

        this.iconPanel.setReadyToPlay(isReadyToPlay);
    }

    void setResign(final boolean isResign) {

        this.iconPanel.setResign(isResign);
    }

    boolean isActivePlayer() {
        return this.isActivePlayer;
    }

    /**
     * Sets the flag that indicates whether this player is active (it is the turn of
     * this player). It also may change the visual indication whether this player is
     * thinking or not.
     *
     * @param isActivePlayer whether this player is the active player
     */
    void setActivePlayer(final boolean isActivePlayer) {
        this.isActivePlayer = isActivePlayer;
        // a refresh is needed to toggle the "is thinking" visual within the UI
        this.refreshHeader();
        setBorder(getPanelBorder(this.isActivePlayer));
        updateIssWidgets(isActivePlayer);
    }

    /**
     * Sets the flag that indicates whether this player is an AI player or not.
     *
     * @param isAIPlayer TRUE, if the player is an AI player
     */
    void setAIPlayer(final boolean isAIPlayer) {

        this.isAIPlayer = isAIPlayer;
    }

    private void updateIssWidgets(final boolean isActivePlayer) {
        if (this.showIssWidgets) {
            if (isActivePlayer) {
                this.clockPanel.setActive();
            } else {
                this.clockPanel.setInactive();
            }
        }
    }

    void setPass(final boolean isPassed) {
        this.playerPassed = isPassed;
        refreshHeader();
    }

    void setDeclarer(final boolean isDeclarer) {
        this.declarer = isDeclarer;
        refreshHeader();
    }

    public String getPlayerName() {

        return this.playerName;
    }

    public void setGeschoben() {
        this.playerGeschoben = true;
        refreshHeader();
    }

    public void setContra() {
        this.playerContra = true;
        refreshHeader();
    }

    public void setRe() {
        this.playerRe = true;
        refreshHeader();
    }
}

File: jskat-gui/src/main/kotlin/org/jskat/gui/javafx/table/AbstractHandPanel.kt
package org.jskat.gui.javafx.table

import javafx.geometry.Insets
import javafx.scene.control.Label
import javafx.scene.layout.*
import javafx.scene.paint.Color
import org.jskat.data.JSkatOptions
import org.jskat.gui.img.JSkatGraphicRepository
import org.jskat.util.*

abstract class AbstractHandPanel(
    private val maxCards: Int,
    private val showIssWidgets: Boolean
) : VBox() {

    var position: Player? = null
        set(value) {
            field = value
            refreshHeader()
        }

    internal val bitmaps = JSkatGraphicRepository.INSTANCE
    internal val strings = JSkatResourceBundle.INSTANCE

    private val header = HBox()
    private val headerLabel = Label(" ")
    private val iconPanel = IconPanel()
    private val clockPanel = ClockPanel()

    var playerName: String? = null
        set(value) {
            field = value
            refreshHeader()
        }

    var bidValue: Int = 0
        set(value) {
            field = value
            refreshHeader()
        }

    var isActivePlayer = false
        set(value) {
            field = value
            refreshHeader()
            background = getPanelBackground(value)
            border = getPanelBorder(value)
            updateIssWidgets(value)
        }

    var isAIPlayer = false
    var playerPassed = false
        set(value) {
            field = value
            refreshHeader()
        }
    var playerGeschoben = false
        set(value) {
            field = value
            refreshHeader()
        }
    var declarer = false
        set(value) {
            field = value
            refreshHeader()
        }
    var playerContra = false
        set(value) {
            field = value
            refreshHeader()
        }
    var playerRe = false
        set(value) {
            field = value
            refreshHeader()
        }

    internal var cardPanel: CardPanel

    init {
        minHeight = 0.0
        minWidth = 0.0

        cardPanel = createCardPanel()
        initPanel()
        children.add(cardPanel)
        setVgrow(cardPanel, Priority.ALWAYS)

        if (JSkatOptions.instance().isCheatDebugMode) {
            showCards()
        }
    }

    protected open fun createCardPanel(): CardPanel {
        return CardPanel(1.0, true)
    }

    private fun initPanel() {
        setVgrow(header, Priority.NEVER)
        header.minHeight = USE_PREF_SIZE
        header.minWidth = 0.0
        headerLabel.minWidth = 0.0
        headerLabel.style = "-fx-text-fill: white;"

        background = getPanelBackground(isActivePlayer)
        border = getPanelBorder(isActivePlayer)

        val headerInsets = if (showIssWidgets) {
            "-fx-padding: 0 5 0 0;"
        } else {
            "-fx-padding: 5;"
        }
        header.style = headerInsets

        header.children.add(headerLabel)
        val blankPanel = HBox()
        HBox.setHgrow(blankPanel, Priority.ALWAYS)
        header.children.add(blankPanel)
        header.children.add(iconPanel)

        if (showIssWidgets) {
            iconPanel.isShowIssWidgets = true
            header.children.add(clockPanel)
        }
        children.add(header)
        VBox.setMargin(cardPanel, Insets(0.0, 5.0, 5.0, 5.0))
    }

    private fun getPanelBackground(isActivePlayer: Boolean): Background {
        return Background(
            BackgroundFill(
                Color.web("#486760"),
                CornerRadii(10.0),
                null
            )
        )
    }

    private fun getPanelBorder(isActivePlayer: Boolean): Border {
        return if (isActivePlayer) {
            Border(
                BorderStroke(
                    Color.rgb(255, 191, 0),
                    BorderStrokeStyle.SOLID,
                    CornerRadii(10.0),
                    BorderWidths(3.0)
                )
            )
        } else {
            Border(
                BorderStroke(
                    Color.BLACK,
                    BorderStrokeStyle.SOLID,
                    CornerRadii(10.0),
                    BorderWidths(3.0)
                )
            )
        }
    }

    private fun refreshHeader() {
        val headerText = StringBuilder()
        headerText.append(playerName).append(": ")

        if (position != null) {
            when (position) {
                Player.FOREHAND -> headerText.append(strings.getString("forehand"))
                Player.MIDDLEHAND -> headerText.append(strings.getString("middlehand"))
                Player.REARHAND -> headerText.append(strings.getString("rearhand"))
                else -> {
                }
            }

            headerText.append(" ${strings.getString("bid")}: ")
            headerText.append(bidValue)

            if (playerPassed || playerGeschoben || playerContra || playerRe) {
                headerText.append(" (")
                var passedGeschobenContraRe = ""
                if (playerPassed) {
                    passedGeschobenContraRe = strings.getString("passed")
                }
                if (playerGeschoben) {
                    passedGeschobenContraRe = strings.getString("geschoben")
                }

                if (passedGeschobenContraRe.isNotEmpty() && (playerContra || playerRe)) {
                    passedGeschobenContraRe += " "
                }

                if (playerContra) {
                    passedGeschobenContraRe += strings.getString("contra")
                }
                if (playerRe) {
                    passedGeschobenContraRe += strings.getString("re")
                }
                headerText.append(passedGeschobenContraRe)
                headerText.append(")")
            }

            if (declarer) {
                headerText.append(" (${strings.getString("declarer")})")
                headerLabel.style = "-fx-font-weight: bold; -fx-text-fill: white;"
            } else {
                headerLabel.style = "-fx-font-weight: normal; -fx-text-fill: white;"
            }
        }

        iconPanel.setThinking(isActivePlayer && isAIPlayer)
        headerLabel.text = headerText.toString()
    }

    fun addCard(newCard: Card) {
        cardPanel.addCard(newCard)
    }

    fun addCards(newCards: CardList) {
        cardPanel.addCards(newCards)
    }

    fun removeCard(cardToRemove: Card) {
        cardPanel.removeCard(cardToRemove)
    }

    fun removeAllCards() {
        cardPanel.clearCards()
    }

    fun clearHandPanel() {
        cardPanel.clearCards()
        bidValue = 0
        playerPassed = false
        playerGeschoben = false
        playerContra = false
        playerRe = false
        declarer = false
        iconPanel.reset()
        refreshHeader()
        isActivePlayer = false
        hideCards()
    }

    open fun hideCards() {
        cardPanel.hideCards()
    }

    fun showCards() {
        cardPanel.showCards()
    }

    fun setSortGameType(newGameType: GameType) {
        cardPanel.setSortType(newGameType)
    }

    fun isHandFull(): Boolean {
        return cardPanel.cardCount == maxCards
    }

    fun setPlayerTime(newTime: Double) {
        clockPanel.setPlayerTime(newTime)
    }

    fun setChatEnabled(isChatEnabled: Boolean) {
        iconPanel.isChatEnabled = isChatEnabled
        iconPanel.refreshIcons()
    }

    fun setReadyToPlay(isReadyToPlay: Boolean) {
        iconPanel.isReadyToPlay = isReadyToPlay
        iconPanel.refreshIcons()
    }

    fun setResign(isResign: Boolean) {
        iconPanel.setResign(isResign)
    }

    private fun updateIssWidgets(isActivePlayer: Boolean) {
        if (showIssWidgets) {
            if (isActivePlayer) {
                clockPanel.setActive()
            } else {
                clockPanel.setInactive()
            }
        }
    }

    fun setContra() {
        playerContra = true
    }

    fun setRe() {
        playerRe = true
    }

    fun setGeschoben() {
        playerGeschoben = true
    }
}
